ffi/src/lib.rs

```diff
@@ -620,30 +620,136 @@ pub unsafe extern "C" fn version(snapshot: *const SnapshotHandle) -> u64 {
     snapshot.version()
 }
 
+/// The `EngineSchemaVisitor` defines a visitor system to allow engines to build their own
+/// representation of a schema from a particular schema within kernel.
+///
+/// The model is list based. When the kernel needs a list, it will ask engine to allocate one of a
+/// particular size. Once allocated the engine returns an `id`, which can be any identifier the
+/// engine wants, and will be passed back to the engine to identify the list in the future.
+///
+/// Every schema element the kernel visits belongs to some list of "sibling" elements. The schema
+/// itself is a list of schema elements, and every complex type (struct, map, array) contains a list
+/// of "child" elements.
+///  1. Before visiting schema or any complex type, the kernel asks the engine to allocate a list to
+///     hold its children
+///  2. When visiting any schema element, the kernel passes its parent's "child list" as the
+///     "sibling list" the element should be appended to:
+///      - For the top-level schema, visit each top-level column, passing the column's name and type
+///      - For a struct, first visit each struct field, passing the field's name, type, nullability,
+///        and metadata
+///      - For a map, visit the key and value, passing its special name ("map_key" or "map_value"),
+///        type, and value nullability (keys are never nullable)
+///      - For a list, visit the element, passing its special name ("array_element"), type, and
+///        nullability
+///  3. When visiting a complex schema element, the kernel also passes the "child list" containing
+///     that element's (already-visited) children.
+///  4. The [`visit_schema`] method returns the id of the list of top-level columns
 // WARNING: the visitor MUST NOT retain internal references to the string slices passed to visitor methods
-// TODO: other types, nullability
+// TODO: struct nullability and field metadata
 #[repr(C)]
 pub struct EngineSchemaVisitor {
-    // opaque state pointer
-    data: *mut c_void,
-    // Creates a new field list, optionally reserving capacity up front
-    make_field_list: extern "C" fn(data: *mut c_void, reserve: usize) -> usize,
+    /// opaque state pointer
+    pub data: *mut c_void,
+    /// Creates a new field list, optionally reserving capacity up front
+    pub make_field_list: extern "C" fn(data: *mut c_void, reserve: usize) -> usize,
+
     // visitor methods that should instantiate and append the appropriate type to the field list
-    visit_struct: extern "C" fn(
+    /// Indicate that the schema contains a `Struct` type. The top level of a Schema is always a
+    /// `Struct`. The children of the `Struct` are in the list identified by `child_list_id`.
```

Comment:
Does this make sense?
```suggestion
    /// `Struct`. The fields of the `Struct` are in the list identified by `child_list_id`.
```